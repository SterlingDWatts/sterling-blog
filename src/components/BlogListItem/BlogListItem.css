.BlogListItem {
  margin: 8px;
}
