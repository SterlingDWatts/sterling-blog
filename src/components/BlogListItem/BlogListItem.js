import React, { Component } from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";

import "./BlogListItem.css";

class BlogListItem extends Component {
  render() {
    const length = Math.floor(this.props.content.length / 300);
    return (
      <li className="BlogListItem">
        <div className="BlogListItem__info">
          <Link
            to={`/blog/${this.props.id}`}
            className="BlogListItem__title_link"
          >
            <h3 className="BlogListItem__title">{this.props.title}</h3>
          </Link>
          <div className="BlogListItem__date_and_length">
            {this.props.date + " Â· " + length + " min read"}
          </div>
          <div className="BlogListItem__author">{this.props.author}</div>
          <div className="BlogListItem__line"></div>
        </div>
        <div className="BlogListItem__pic_holder">
          <Link to={`/blog/${this.props.id}`}>
            <div
              className="BlogListItem__pic"
              style={{
                backgroundImage: "url('" + this.props.squarePic + "')"
              }}
            ></div>
          </Link>
        </div>
      </li>
    );
  }
}

BlogListItem.propTypes = {
  id: PropTypes.number.isRequired,
  views: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  content: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  squarePic: PropTypes.string.isRequired,
  date: (props, propName, componentName) => {
    const prop = props[propName];

    if (!prop) {
      return new Error(`${propName} is required in ${componentName}`);
    }

    if (typeof prop != "string") {
      return new Error(
        `Invalid prop, ${propName} is expected to be a string in ${componentName}. ${typeof prop} found.`
      );
    }

    if (!new Date(prop).getTime()) {
      return new Error(
        `Invalid prop, ${propName} is expected to be a date in ${componentName}`
      );
    }
  }
};

export default BlogListItem;
