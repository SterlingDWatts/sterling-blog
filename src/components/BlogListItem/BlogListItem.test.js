import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import BlogListItem from "./BlogListItem";

describe("BlogListItem Component", () => {
  const blog = {
    id: 1,
    title: "test title",
    content: "test content",
    author: { firs_name: "test", last_name: "author" },
    picture: "https://picsum.photos/seed/three/900/510",
    views: 8,
    date: new Date("December 22, 2019")
  };
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <BrowserRouter>
        <BlogListItem {...blog} />
      </BrowserRouter>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders the UI as expected", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <BlogListItem {...blog} />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
