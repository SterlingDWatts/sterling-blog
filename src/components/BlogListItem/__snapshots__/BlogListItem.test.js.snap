// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogListItem Component renders the UI as expected 1`] = `
<li
  className="BlogListItem"
>
  <div
    className="BlogListItem__info"
  >
    <a
      className="BlogListItem__title_link"
      href="/blog/1"
      onClick={[Function]}
    >
      <h3
        className="BlogListItem__title"
      >
        test title
      </h3>
    </a>
    <div
      className="BlogListItem__date_and_length"
    >
      December 22, 2019 Â· 0 min read
    </div>
    <div
      className="BlogListItem__author"
    >
      test author
    </div>
    <div
      className="BlogListItem__line"
    />
  </div>
  <div
    className="BlogListItem__pic_holder"
  >
    <a
      href="/blog/1"
      onClick={[Function]}
    >
      <div
        className="BlogListItem__pic"
        style={
          Object {
            "backgroundImage": "url('https://picsum.photos/seed/three/255/255')",
          }
        }
      />
    </a>
  </div>
</li>
`;
